typing/rec_check.ml

```diff
@@ -774,18 +774,77 @@ let rec expression : Typedtree.expression -> term_judg =
         path pth << Dereference;
         list field fields << Dereference;
       ]
-    | Texp_function { cases } ->
+    | Texp_function (params, body) ->
       (*
-         (Gi; _ |- pi -> ei : m[Delay])^i
-         --------------------------------------
-         sum(Gi)^i |- function (pi -> ei)^i : m
+         Function_cases:
 
-         Contrarily to match, the value that is pattern-matched
-         is bound locally, so the pattern modes do not influence
+         (Gi; _ |- pi -> ei : m[Delay])^i
+         (Hj    |-{def}  Pj : m[Delay])^j
+         G  := sum(Gi)^i
+         H  := sum(Hj)^j
+         ps := sum(pat(Pj))^j
+         ---------------------------------------------------------
+         sum(G, H) - ps |- fun (Pj)^j -> function (pi -> ei)^i : m
+
+         Function_body:
+
+         G |- e : m[Delay]
+         (Hj |-{def}  Pj : m[Delay])^j
+         H  := sum(Hj)^j
+         ps := sum(pat(Pj))^j
+         -------------------------------------
+         sum(G, H) - ps |- fun (Pj)^j -> e : m
+
+         Contrarily to match, the values that are pattern-matched
+         are bound locally, so the pattern modes do not influence
          the final environment.
```

Comment:
Comments from review:
  * The "contrarily to match" comment needs to be associated more strongly with the "G_i; _" premise -- add an asterisk?
  * `G + H` is more standard way (in this file) to do `sum(G, H)`.